### ISSUE TITLE
Tracking issue for `const fn` `type_name`

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#63084]
This is a tracking issue for making and stabilizing `type_name` as `const fn`. It is not clear whether this is sound. Needs some T-lang discussion probably, too.

Steps needed:

* [ ] Implementation (essentially add `const` and `#[rustc_const_unstable(feature = "const_type_name")` to the function and add tests showing that it works at compile-time.
* [ ] Stabilization
	* Note that stabilization isn't happening any time soon. This function can change its output between rustc compilations and allows breaking referential transparency. Doing so at compile-time has not been discussed fully and has various points that are problematic.

**Comment by RalfJung** [comment]
> This function can change its output between rustc compilations

Can the output change between compiling a library crate, and a binary crate using that library? Or only when switching rustc versions?

**Comment by oli-obk** [comment]
> Can the output change between compiling a library crate, and a binary crate using that library?

No, we're using global paths now.

> Or only when switching rustc versions?

Yes, there needs to be a change in rustc that causes a change in output.

**Comment by TankhouseAle** [comment]
I might take a look at this. (I implemented the support for _using_ `type_name` inside of `const fn`s, if you don't remember me, haha.)

**Comment by TankhouseAle** [comment]
Sorry it took me a bit to get to this. [Just opened a PR.](https://github.com/rust-lang/rust/pull/63123)

**Comment by lfairy** [comment]
How does this break referential transparency? I'd like to see that elaborated upon.

**Comment by TankhouseAle** [comment]
I might be wrong, but I think the first "checkbox" can be checked since my PR was merged?

@lfairy:

It doesn't IMO. Some people however seem to think that despite the fact Rust has fully reified generics, it's somehow a "dangerous" concept for your "average Joe" to be able to access the (100%-guaranteed-correct) name of a generic parameter.

**Comment by korken89** [comment]
Hi, is there a stabilization issue for this `const`-ification?

**Comment by oli-obk** [comment]
This is the tracking issue, so it will get closed on stabilization. I don't know what the next steps for stabilizing it are though. The discussion around referential transparency needs to be had (although with specialization it'd become obsolete, because then we'd get this feature anyway, but in a less convenient way). I think posting use cases would be very helpful for the discussion. Since this works with feature gates, you can always show a use case with nightly.

**Comment by korken89** [comment]
Hi, certainly I can write about a use case that I am working on!

I'm part of the Embedded WG where I am currently working on an instrumentation/logging tool for embedded targets. Here, as you probably know, we have strict requirements on code size as the micro controllers have very little memory.
And a huge hog of memory is when strings are placed in memory.

To this end I place formating strings and (I want) type strings in an `INFO` linker section. That is the following code as an example:

```rust
// test1 is some variable
mylog::log!("Look what I got: {}", &test1);

//
// Expands to:
//

const fn get_type_str<T>(_: &T) -> &'static str {
    // type_name needs to be a const fn for this to work
    core::any::type_name::<T>()
}

// ugly hack to tranform `&'static str` -> `[u8; str.len()]`
union Transmute<T: Copy, U: Copy> {
    from: T,
    to: U,
}

const FMT: &'static str = "Look what I got: {}";
const TYP: &'static str = get_type_str(&test1);

// Transform string into byte arrays at compile time so 
// they can be placed in a specific memory region

// Formating string placed in a specific linker section
#[link_section = ".some_info_section"]
static F: [u8; FMT.as_bytes().len()] = unsafe {
    *Transmute::<*const [u8; FMT.len()], &[u8; FMT.as_bytes().len()]> {
        from: FMT.as_ptr() as *const [u8; FMT.as_bytes().len()],
    }
    .to
};

// Type string placed in a specific linker section
#[link_section = ".some_info_section"]
static T: [u8; TYP.as_bytes().len()] = unsafe {
    *Transmute::<*const [u8; TYP.len()], &[u8; TYP.as_bytes().len()]> {
        from: TYP.as_ptr() as *const [u8; TYP.as_bytes().len()],
    }
    .to
};

// Here we send where the strings are stored in the ELF + the data to the host 
write_frame_to_host(&F, &T, &test1)
```

Link script:

```
SECTIONS {
  .some_info_section (INFO) :
  {
    *(.some_info_section .some_info_section.*);
  }
}
```

By doing this the strings are available in the generated ELF but not loaded onto the target, and the type can be reconstructed through DWARF with help of the type string.
And with complex types the space used for the type is getting very large for example any type that uses `GenericArray` (which we use extensively).

Currently the string is placed in `.rodata` which means that the compile fully knows that this is a static string, and I made an issue on how to control where strings are placed which can use a trick outlined in here: https://github.com/rust-lang/rust/issues/70239
However this trick only works if the string is available in const context.

This is why I am asking about stabilization here.

**Comment by That3Percent** [comment]
I would also like to get this stabilized and can present a use-case.

I'm working on [firestorm](https://github.com/That3Percent/firestorm): a low overhead intrusive flame graph profiler. I created `firestorm` because existing solutions have too much overhead to be useful for [Tree-Buf](https://github.com/That3Percent/tree-buf): the self-describing serialization system that produces files smaller than GZip faster than uncompressed formats.

Part of keeping the overhead of `firestorm` low is to delay string formatting/allocation until outside of the profiled region. So, there is an enum that allows for arbitrary amounts of `&'static str` to be concatenated. In order to keep the amount of data writing to a minimum during the execution of the profiled region, we only write `&'static EventData` so that `EventData` can be large. Supporting concatenation of fn name and struct name for a method is only possible if `type_name` is a `const fn`.

TLDR: I need this to be `const fn` for performance.

**Comment by That3Percent** [comment]
To elaborate here's code I would like to have be able to compile:

```rust
#[macro_export]
macro_rules! profile_method {
    ($($t:tt)*) => {
        let _firestorm_method_guard = {
            const FIRESTORM_STRUCT_NAME: &'static str = ::std::any::type_name::<Self>();
            let event_data: &'static _ = &$crate::internal::EventData::Start(
                $crate::internal::Start::Method {
                    signature: stringify!($($t)*),
                    typ: FIRESTORM_STRUCT_NAME,
                }
            );
            $crate::internal::start(event_data);
            $crate::internal::SpanGuard
        };
    };
}
```

It seems there are at least 2 problems here though. First, is that `type_name` is not a const fn. The second is that the compiler complains about the use of `Self` as the generic for the const fn. I do not really understand why that's a problem when each monomorphized function could have its own value here.

**Comment by That3Percent** [comment]
@oli-obk Yes, you are right. Both issues would need to be resolved to support my use-case in the way that I would like.

**Comment by oli-obk** [comment]
I found a bug with `type_name`: `type_name` works in array lengths of arrays used in associated consts, even if it is `type_name::<Self>`. This doesn't even work for `size_of`. The bug is that `type_name` just returns `_` in that case:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4404a9b3635b4bf5047c2493ac9a5082

If you comment out the first panic, you get the correct type name.

**Comment by RalfJung** [comment]
> I found a bug with type_name: type_name works in array lengths of arrays used in associated consts, even if it is type_name::<Self>. This doesn't even work for size_of. The bug is that type_name just returns _ in that case:

This seems to be fixed now?

**Comment by korken89** [comment]
Has there been any progress on the stabilization of `type_name`?

**Comment by korken89** [comment]
Ping again on the status of stabilization?

**Comment by Hoverbear** [comment]
I was wanting to use this today to create a type like so:

```rust
pub trait WithTypeIds {
    const ITEM_ID: TypeId;
    const OPTION_ID: Option<TypeId>;
    const VEC_ID: Option<TypeId>;
    const VEC_OPTION_ID: Option<TypeId>;
    const OPTION_VEC_ID: Option<TypeId>;
    const OPTION_VEC_OPTION_ID: Option<TypeId>;
}
```

So that later, on instantiated types, I could access various typeids of these containers, and use it to map Rust types to equivalent SQL types. This is for https://github.com/zombodb/pgx.

There have been no reported issues for several releases, what steps remain towards stabilizing this?

**Comment by RalfJung** [comment]
`type_id` (https://github.com/rust-lang/rust/issues/77125) has a bunch of issues which is why its stabilization actually got reverted; not sure if those issues also apply to `type_name`.

**Comment by Hoverbear** [comment]
Would you believe I clicked the wrong link in Rustdoc? Sorry.

**Comment by TheNeikos** [comment]
Is there any outlook on stabilization/progress towards it?

**Comment by oli-obk** [comment]
There's a rendering bug and a soundness bug that we need to fix before we can stabilize. Both are linked from the main post

**Comment by terrarier2111** [comment]
> There's a rendering bug and a soundness bug that we need to fix before we can stabilize. Both are linked from the main post

The soundness bug seems to have been fixed by now, and all other linked issues also seem to be fixed, is there any particular roadblock remaining?

**Comment by Bromeon** [comment]
With the recent `const {}` stabilization, this would prove very useful in static assertion messages.

Are there more known roadblocks to `const fn type_name<T>()`?

**Comment by tgross35** [comment]
Started a discussion https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Const.20.60TypeId.3A.3Aof.60.2C.20.60TypeId.3A.3Amatches.60.2C.20.60type_name.60.2C

**Comment by ickk** [comment]
What were the fruits of that discussion? What blockers, if any, were named? 

Can this tracking issue be updated with any outstanding concerns

**Comment by RalfJung** [comment]
The main concern that got raised is that if we give people type_name but not type_id, they will start using type_name comparison to check if two types are equal. The docs explicitly say not to do that, but it'll happen anyway when the proper alternative (type_id comparison) is simply not available.

@oli-obk 
> Note that stabilization isn't happening any time soon. This function can change its output between rustc compilations and allows breaking referential transparency. Doing so at compile-time has not been discussed fully and has various points that are problematic.

Do you remember what this is about? In which sense does this function break referential transparency? We do have to ensure that the result is consistent across all crates in a crate graph but I assume that is already the case? Or are there `-C`/`-Z` flags that can change the behavior of this function?

**Comment by oli-obk** [comment]
`-Zverbose` used to change the output. We've since switched to `-Zverbose-types`, but I do not know if that still affects the output

**Comment by RalfJung** [comment]
Hm, that seems like a potential soundness concern then, doesn't it? The same const could produce different results when it is sometimes evaluated with that flag and sometimes without.

Is there any more systematic way we can avoid accidentally having const-eval depend on such flags? This is a really easy mistake to make and it'll hardly ever be detected... but it is a soundness issue.

**Comment by ickk** [comment]
@RalfJung 
> The main concern that got raised is that if we give people type_name but not type_id, they will start using type_name comparison to check if two types are equal. The docs explicitly say not to do that, but it'll happen anyway when the proper alternative (type_id comparison) is simply not available.

If the `type_id` solution is still a long way out, and if the main concern is misuse of the api: would it be possible to add a lint against that misuse to try to control for it?

**Comment by RalfJung** [comment]
No, it is not unblocked. type_id *comparison* is still not possible, and that is the actual blocker here as explained above.

**Comment by korken89** [comment]
Thanks for clarifying!

**Comment by oli-obk** [comment]
your LLM analysis is inconsistent, unnecessarily contains things irrelevant to further discussion, contains trivial stabilization PR instructions, and idk what you mean with lang/libs confirming a blocker. Please do not post LLM generated analyses, so far we have not seen a correct one
